### PR TITLE
data sampling refactor + online eval

### PR DESCRIPTION
**Pull request by katerakelly**
submitting as one PR because disentangling some of the commits was going to require a lot of re-writing

### CODE REVIEW

File: rlkit/envs/point_robot.py [hunk 1]
```diff
@@ -8,13 +8,17 @@
...
+    tasks (aka goals) are positions on the plane
+
+     - tasks sampled from unit square
+     - reward is L2 distance
     """
 
     def __init__(self, randomize_tasks=False, n_tasks=2):
 
         if randomize_tasks:
+            np.random.seed(1337)
```
Comment (azhou42) [inline rlkit/envs/point_robot.py:18]: but why

Comment (katerakelly) [reply rlkit/envs/point_robot.py:18]: my old collaborator used it for everything and I picked up the habit haha XD

File: rlkit/torch/sac/proto.py [hunk 2]
```diff
@@ -16,26 +16,42 @@ def _product_of_gaussians(mus, sigmas_squared):
...
 class ProtoAgent(nn.Module):
 
     def __init__(self,
+                 env,
                  latent_dim,
                  nets,
                  **kwargs
     ):
         super().__init__()
+        self.env = env
```
Comment (azhou42) [inline rlkit/torch/sac/proto.py]: I don't feel too strongly, but I don't think the agent should have access to the env here. If all we're doing is calling sparsify reward, it's probably cleaner to just store it as an extra field in the replay buffer. It also just seems pretty hacky to me that the sparse reward is just a function of the original reward.

Comment (katerakelly) [reply rlkit/torch/sac/proto.py]: I agree - we can do what I did when I was trying to make the MAESN envs work - stuff the sparse reward into `env_infos` (adding another field to the replay buffer is a pain).

File: rlkit/torch/sac/proto.py [hunk 3]
```diff
@@ -54,90 +70,116 @@ def __init__(self,
...
-            for i in range(num_updates):
-                num_z += 1
-                z += (new_z[i][None] - z) / num_z
-        if self.use_ib:
-            z = self.information_bottleneck(z)
+            num_updates = up_params.size(1)
+            # only have one task here
+            up_params = up_params[0] # batch x feat
+            curr_mu = self.z_means[0]
+            curr_ss = self.z_vars[0]
```
Comment (azhou42) [inline rlkit/torch/sac/proto.py:165]: will z_vars[0] exist if we're not using ib? this might crash

Comment (katerakelly) [reply rlkit/torch/sac/proto.py:165]: yes, in the deterministic case, `z_vars` is set to zero (see line 77)

File: rlkit/torch/sac/proto.py [hunk 4]
```diff
@@ -54,90 +70,116 @@ def __init__(self,
...
+                new_natural = torch.cat(_canonical_to_natural(up_mu, up_ss), dim=1) # batch x feat
+                new_natural = torch.sum(new_natural, dim=0)
+                natural_z += new_natural
+                m, s  = _natural_to_canonical(natural_z[:self.latent_dim], natural_z[self.latent_dim:]) # feat
+                self.z_means = m[None]
+                self.z_vars = s[None]
+            else:
+                for i in range(num_updates):
+                    num_z += 1
+                    curr_mu += (up_params[i] - curr_mu) / num_z
```
Comment (azhou42) [inline rlkit/torch/sac/proto.py:179]: this computation seems bugged. if it's actually the mean, then it should be 
(x + (n-1) * mu) / n?

File: rlkit/torch/torch_rl_algorithm.py [hunk 5]
```diff
@@ -46,187 +46,130 @@ def get_batch(self, idx=None):
...
-        ### train tasks
-        dprint('evaluating on {} train tasks'.format(len(self.train_tasks)))
-        train_avg_returns = []
-        for idx in self.train_tasks:
-            dprint('task {} encoder RB size'.format(idx), self.enc_replay_buffer.task_buffers[idx]._size)
-            paths = self.collect_paths(idx, epoch, eval_task=False)
-            train_avg_returns.append(eval_util.get_average_returns(paths))
+        ### sample trajectories from prior for vis
+        prior_paths = []
+        for _ in range(10):
```
Comment (azhou42) [inline rlkit/torch/torch_rl_algorithm.py]: make 10 a parameter?

File: rlkit/torch/sac/proto.py (hunk 4 above)
Comment (katerakelly) [reply rlkit/torch/sac/proto.py:179]: it's equivalent. written this way it's just factorized into the sum of the old value and the new part

File: rlkit/torch/torch_rl_algorithm.py (hunk 5 above)
Comment (katerakelly) [reply rlkit/torch/torch_rl_algorithm.py]: or better yet I think this should go into `debug` mode since aside from visualization it's just a waste of computation

Comment (katerakelly) [reply rlkit/torch/torch_rl_algorithm.py]: see `258b316`...it's hard-coded to collecting 100 trajs, but that seems ok to me

Comment (azhou42) [reply rlkit/torch/torch_rl_algorithm.py]: ya seems fine if its just for debugging

File: examples/default.py [hunk 6]
```diff
@@ -36,10 +36,11 @@ def make_variant(max_path_length):
             reparameterize=True, # should always be True
             kl_lambda=.1, # weight on KL divergence term in encoder loss
             use_information_bottleneck=True, # False makes latent context deterministic
+            # train embedding source should choose from {'initial_pool', 'online_exploration_trajectories', 'online_on_policy_trajectories'}
             train_embedding_source='online_exploration_trajectories',
-            # embedding_source should be chosen from
-            # {'initial_pool', 'online_exploration_trajectories', 'online_on_policy_trajectories'}
-            eval_embedding_source='online_exploration_trajectories',
+            resample_z='trajectory', # how often to resample z at test-time
+            resample_z_train=max_path_length, # how often to resample z at training time
```
Comment (azhou42) [inline examples/default.py]: Seems weird to have resample_z take value 'trajectory', and resample_z_train=max path length.
I think we should maintain a consistent structure for controlling how embeddings are done at train and test.

File: rlkit/core/rl_algorithm.py [hunk 7]
```diff
@@ -189,31 +185,17 @@ def train(self):
                 if self.train_embedding_source == 'initial_pool':
                     # embeddings are computed using only the initial pool of data
                     # sample data from posterior to train RL algorithm
-                    self.collect_data_from_task_posterior(idx=idx,
-                                                          num_samples=self.num_steps_per_task,
-                                                          add_to_enc_buffer=False)
-                elif self.train_embedding_source == 'posterior_only':
-                    self.collect_data_from_task_posterior(idx=idx, num_samples=self.num_steps_per_task, eval_task=False,
-                                                          add_to_enc_buffer=True)
+                    self.collect_data(self.num_steps_per_task, self.resample_z_train, np.inf, self.embedding_batch_size, add_to_enc_buffer=False)
```
Comment (azhou42) [inline rlkit/core/rl_algorithm.py]: I think it's also okay to just get rid of this option altogether. I don't see when someone would use it at this point.

File: rlkit/core/rl_algorithm.py [hunk 8]
```diff
@@ -224,7 +206,7 @@ def train(self):
                 self._n_train_steps_total += 1
             gt.stamp('train')
 
-            #self.training_mode(False)
+            self.training_mode(False)
```
Comment (azhou42) [inline rlkit/core/rl_algorithm.py:204]: Do we actually use this flag now?

File: rlkit/core/rl_algorithm.py [hunk 9]
```diff
@@ -238,74 +220,31 @@ def pretrain(self):
...
+        '''
+        # makes no sense to update the posterior without resampling z
+        assert resample_z_rate <= update_posterior_rate
+        # start from the prior
+        self.reset_posterior()
+        counter = 0
+        for _ in range(num_samples // resample_z_rate):
+            self.collect_transitions(self.policy, num_samples=resample_z_rate, add_to_enc_buffer=add_to_enc_buffer)
+            counter += resample_z_rate
+            if counter % update_posterior_rate == 0:
```
Comment (azhou42) [inline rlkit/core/rl_algorithm.py]: If for some reason resample_z_rate does not divide update_posterior_rate, this doesn't work as it is probably intended. Also, given that we reset the envs between switching tasks, maybe collect transitions should just collect complete trajectories until we exceed num_samples (returning the # of smaples it consumed), and then update posterior if stepcount >= update_count * posterior rate.

File: rlkit/samplers/util.py [hunk 10]
```diff
@@ -51,6 +52,8 @@ def rollout(env, agent, max_path_length=np.inf, animated=False, is_online=False)
         o = next_o
         if animated:
             env.render()
+    if resample == 'trajectory':
```
Comment (azhou42) [inline rlkit/samplers/util.py]: personally don't think this should belong in the rollout method (should just be called in the while loop in obtain_samples)

File: examples/default.py (hunk 6 above)
Comment (katerakelly) [reply examples/default.py]: I agree...my original plan was to have all resample at any arbitrary rate. I'm not sure how to handle episodes that terminate before `max_path_length` though. Should we have a special argument value that means "trajectory" however long that happens to be? Or we could go the other way and have just transitions and multiples of trajectories

File: rlkit/core/rl_algorithm.py (hunk 7 above)
Comment (katerakelly) [reply rlkit/core/rl_algorithm.py]: do you think the same is true of `initial_pool`?

File: examples/default.py (hunk 6 above)
Comment (azhou42) [reply examples/default.py]: We could just have it consistently resample by trajectory instead of max_path_length?

File: rlkit/core/rl_algorithm.py (hunk 8 above)
Comment (katerakelly) [reply rlkit/core/rl_algorithm.py:204]: ah I meant to make this a separate commit to highlight it....it breaks the LSTM for reasons unknown to me but I think it's unsafe to omit it unless we have to, and I don't think `master` needs to support the LSTM

File: rlkit/core/rl_algorithm.py (hunk 7 above)
Comment (azhou42) [reply rlkit/core/rl_algorithm.py]: This comment was addressed towards the initial_pool option.

File: rlkit/samplers/util.py (hunk 10 above)
Comment (katerakelly) [reply rlkit/samplers/util.py]: I disagree because I think all the resampling should happen in the same function, and transition-level resampling has to happen here. Though if we generalize it to resample after any number of transitions, we'll run into this problem again...

Comment (azhou42) [reply rlkit/samplers/util.py]: My reasoning on keeping it out is that rollout shouldn't care about what happens after the rollout, but I'm fine either way. I'm against resampling after arbitrary numbers of transitions. I think it should just be sample whole trajectories until we reach a threshold to resample.

File: rlkit/core/rl_algorithm.py (hunk 7 above)
Comment (katerakelly) [reply rlkit/core/rl_algorithm.py]: see `9161b6b`

File: rlkit/core/rl_algorithm.py (hunk 9 above)
Comment (katerakelly) [reply rlkit/core/rl_algorithm.py]: see `81f27a8`

File: rlkit/core/rl_algorithm.py [hunk 11]
```diff
@@ -217,41 +217,39 @@ def pretrain(self):
...
         # start from the prior
         self.reset_posterior()
-        counter = 0
+        num_collected = 0
+        update_count = 1
         for _ in range(num_samples // resample_z_rate):
-            self.collect_transitions(self.policy, num_samples=resample_z_rate, add_to_enc_buffer=add_to_enc_buffer)
-            counter += resample_z_rate
-            if counter % update_posterior_rate == 0:
+            num_collected += self.collect_transitions(self.policy, num_samples=resample_z_rate, add_to_enc_buffer=add_to_enc_buffer)
```
Comment (azhou42) [inline rlkit/core/rl_algorithm.py]: collect_transitions should be renamed to collect_trajectories

File: rlkit/core/rl_algorithm.py [hunk 12]
```diff
@@ -217,41 +217,39 @@ def pretrain(self):
...
+        collect complete trajectories until the number of collected transitions >= num_samples
         '''
-        for _ in range(num_samples):
+        num_collected = 0
+        while num_collected < num_samples:
             action, agent_info = self._get_action_and_info(agent, self.train_obs)
             if self.render:
                 self.env.render()
-            next_ob, raw_reward, terminal, env_info = (
+            next_ob, reward, terminal, env_info = (
```
Comment (azhou42) [inline rlkit/core/rl_algorithm.py:238]: do you know why it was originally raw reward here?

File: rlkit/core/rl_algorithm.py [hunk 13]
```diff
@@ -265,14 +263,16 @@ def collect_transitions(self, agent, num_samples=1, add_to_enc_buffer=True):
                 agent_info=agent_info,
                 env_info=env_info,
             )
-            if terminal or len(self._current_path_builder) >= self.max_path_length:
+            num_collected += 1
+            if terminal:
```
Comment (azhou42) [inline rlkit/core/rl_algorithm.py]: are you sure we don't need to check the max path length on the algorithm side?

File: rlkit/core/rl_algorithm.py (hunk 12 above)
Comment (katerakelly) [reply rlkit/core/rl_algorithm.py:238]: yes because reward scaling used to be done here -_-

File: rlkit/core/rl_algorithm.py (hunk 13 above)
Comment (katerakelly) [reply rlkit/core/rl_algorithm.py]: you're right, I accidentally deleted that! fixed in `de1b5c3`

File: rlkit/core/rl_algorithm.py (hunk 11 above)
Comment (katerakelly) [reply rlkit/core/rl_algorithm.py]: see `de1b5c3`

File: rlkit/core/rl_algorithm.py (hunk 12 above)
Comment (azhou42) [reply rlkit/core/rl_algorithm.py:238]: 😱

File: examples/default.py (hunk 6 above)
Comment (katerakelly) [reply examples/default.py]: done in `de1b5c3`

File: rlkit/samplers/util.py (hunk 10 above)
Comment (katerakelly) [reply rlkit/samplers/util.py]: ok I think we should just sample by trajectory at test time to keep things consistent, so I'll move this out of the rollout function

Comment (katerakelly) [reply rlkit/samplers/util.py]: see `da9217c`

File: rlkit/torch/sac/proto.py (hunk 2 above)
Comment (katerakelly) [reply rlkit/torch/sac/proto.py]: see `9aaf099`

File: rlkit/core/rl_algorithm.py [hunk 14]
```diff
@@ -175,45 +174,23 @@ def train(self):
                 for idx in self.train_tasks:
                     self.task_idx = idx
                     self.env.reset_task(idx)
-                    self.collect_data_sampling_from_prior(num_samples=self.num_initial_steps,
-                                                          resample_z_every_n=self.max_path_length,
-                                                          eval_task=False)
-
+                    self.collect_data(self.policy, self.num_initial_steps, 1, np.inf)
```
Comment (katerakelly) [inline rlkit/core/rl_algorithm.py:177]: I kind of dislike using `np.inf` to get samples from the prior only, but not sure what the cleaner way should look like?

Comment (azhou42) [reply rlkit/core/rl_algorithm.py:177]: I t hink np inf is fine.

Comment (katerakelly) [reply rlkit/core/rl_algorithm.py:177]: let's roll with it for now

**Comment by katerakelly** [comment]
remaining issues: online eval does not give the same results as `master` branch for cheetah-vel

**Comment by katerakelly** [comment]
final status: online latent context computation doesn't quite match with batch computation - probably a numerical issue - so we will close this PR as is and revert back to encoding context batch-wise in a future PR.
